File: hw8/p9-materials/template.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <array>
#include <cassert>

struct Action {
  Action() = default;

  Action(int x, int y, int t) : x(x), y(y), t(t) {
  }

  int x, y, t;
};

struct Task {
  void Read(const std::string& filename) {
    std::ifstream fin(filename);
    assert(fin.is_open());
    Read(fin);
  }

  void Read(std::istream& in = std::cin) {
    in >> test;
    in >> n >> m >> wmax;
    for (int i = 0; i < 5; i++) {
      in >> p[i];
    }
  }

  // Returns -1 if sequence of actions is invalid or too big
  int Evaluate(const std::vector<Action>& solution) {
    if (solution.size() > 500000) {
      return -1;
    }
    std::vector<std::vector<int>> a(n, std::vector<int>(m, -1));
    for (const auto& action : solution) {
      if (action.x <= 0 || action.x > n || action.y <= 0 || action.y > m) {
        return -1;
      }
      if (action.t < 0 || action.t > 4) {
        return -1;
      }
      std::vector<int> used(action.t);
      constexpr std::pair<int, int> dirs[4] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
      for (const auto& dir : dirs) {
        int nx = action.x - 1 + dir.first;
        int ny = action.y - 1 + dir.second;
        if (nx < 0 || nx >= n) {
          continue;
        }
        if (ny < 0 || ny >= m) {
          continue;
        }
        if (a[nx][ny] == -1 || a[nx][ny] >= action.t) {
          continue;
        }
        used[a[nx][ny]] = 1;
      }
      if (std::find(used.begin(), used.end(), 0) != used.end()) {
        return -1;
      }
      a[action.x - 1][action.y - 1] = action.t;
    }
    int score = 0;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (a[i][j] != -1) {
          score += p[a[i][j]];
        }
      }
    }
    return score;
  }

  double Score(int w) {
    if (w <= 0) {
      return 0;
    }
    if (w >= wmax) {
      return 10;
    }
    return (10. * w / wmax) * (1. * w / wmax);
  }

  // Prints solution and returns it's points on test
  double PrintSolution(const std::vector<Action>& solution, const std::string& filename) {
    std::ofstream fout(filename);
    assert(fout.is_open());
    return PrintSolution(solution, fout);
  }

  // Prints solution and returns it's points on test
  double PrintSolution(const std::vector<Action>& solution, std::ostream& out = std::cout) {
    int w = Evaluate(solution);
    out << w << " " << solution.size() << "\n";
    for (const auto& action : solution) {
      out << action.x << " " << action.y << " " << action.t << "\n";
    }
    return Score(w);
  }

  int test{};
  int n{};
  int m{};
  int wmax{};
  std::array<int, 5> p{};
};

int main(int argc, char* argv[]) {
  double total_points = 0;
  for (int test = 1; test <= 10; test++) {
    Task task;
    task.Read("tests/input" + std::to_string(test) + ".txt");
    std::vector<Action> solution;
    for (int i = 0; i < task.n; i++) {
      for (int j = 0; j < task.m; j++) {
        solution.emplace_back(i + 1, j + 1, 0);
      }
    }
    double test_points = task.PrintSolution(solution, "output" + std::to_string(test) + ".txt");
    printf("Test %d: %.4f points\n", test, test_points);
    total_points += test_points;
  }
  printf("Total: %.4f points\n", total_points);
  return 0;
}

File: hw8/materials_p9/template.pas
program AntineutrinoRocket;

type
  TAction = record
    x: longint;
    y: longint;
    t: longint;
  end;

  TTask = record
    test: longint;
    kMaxN: longint;
    kMaxM: longint;
    wmax: longint;
    p: array [0 .. 4] of longint;
  end;

procedure ReadData(var task: TTask; var f: text);
var
  i: longint;
begin
  ReadLn(f, task.test);
  ReadLn(f, task.kMaxN, task.kMaxM, task.wmax);
  for i := 0 to 4 do begin
    Read(f, task.p[i]);
  end;
  ReadLn(f);
end;

procedure ReadData(var task: TTask);
begin
  ReadData(task, input);
end;

procedure ReadData(var task: TTask; const filename: ansistring);
var
  f: text;
begin
  Assign(f, filename);
  Reset(f);
  ReadData(task, f);
  Close(f);
end;

// Returns -1 if sequence of actions is invalid or too big
function Evalulate(const task: TTask; const solution: array of TAction;
                   len: longint): longint;
const
  dx: array [0 .. 3] of longint = (1, -1, 0, 0);
  dy: array [0 .. 3] of longint = (0, 0, 1, -1);
var
  a: array of array of longint;
  i, j: longint;
  used: array [0 .. 4] of boolean;
  action: TAction;
  nx, ny: longint;
  score: longint = 0;
begin
  if len > 500000 then begin
    exit(-1);
  end;
  SetLength(a, task.kMaxN);
  for i := 0 to task.kMaxN - 1 do begin
    SetLength(a[i], task.kMaxM);
    for j := 0 to task.kMaxM - 1 do begin
      a[i, j] := -1;
    end;
  end;
  for i := 0 to len - 1 do begin
    action := solution[i];
    if (action.x <= 0) or (action.x > task.kMaxN) or
       (action.y <= 0) or (action.x > task.kMaxM) or
       (not (action.t in [0 .. 4])) then begin
      exit(-1);
    end;
    for j := 0 to 4 do begin
      used[j] := false;
    end;
    for j := 0 to 3 do begin
      nx := action.x - 1 + dx[j];
      ny := action.y - 1 + dy[j];
      if (nx < 0) or (nx >= task.kMaxN) or (ny < 0) or (ny >= task.kMaxM) then begin
        continue;
      end;
      if a[nx, ny] < 0 then begin
        continue;
      end;
      used[a[nx, ny]] := true;
    end;
    for j := 0 to action.t - 1 do begin
      if not used[j] then begin
        exit(-1);
      end;
    end;
    a[action.x - 1, action.y - 1] := action.t;
  end;
  for i := 0 to task.kMaxN - 1 do begin
    for j := 0 to task.kMaxM - 1 do begin
      if a[i, j] >= 0 then begin
        Inc(score, task.p[a[i, j]]);
      end;
    end;
  end;
  Evalulate := score;
end;

function Score(const task: TTask; w: longint): double;
begin
  if w <= 0 then begin
    exit(0.0);
  end;
  if w >= task.wmax then begin
    exit(10.0);
  end;
  Score := 10.0 * sqr(double(w) / task.wmax);
end;

// Prints solution and returns it's points on test
function PrintSolution(const task: TTask; const solution: array of TAction;
                       len: longint; var f: text): double;
var
  w, i: longint;
  action: TAction;
begin
  w := Evalulate(task, solution, len);
  writeLn(f, w, ' ', len);
  for i := 0 to len - 1 do begin
    action := solution[i];
    writeLn(f, action.x, ' ', action.y, ' ', action.t);
  end;
  PrintSolution := Score(task, w);
end;

// Prints solution and returns it's points on test
function PrintSolution(const task: TTask; const solution: array of TAction;
                       len: longint): double;
begin
  PrintSolution := PrintSolution(task, solution, len, output);
end;

// Prints solution and returns it's points on test
function PrintSolution(const task: TTask; const solution: array of TAction;
                       len: longint; const filename: ansistring): double;
var
  f: text;
begin
  Assign(f, filename);
  Rewrite(f);
  PrintSolution := PrintSolution(task, solution, len, f);
  Close(f);
end;

var
  totalPoints: double = 0;
  test: longint;
  inputName, outputName: ansistring;
  task: TTask;
  solution: array [0 .. 499999] of TAction;
  len: longint;
  i, j: longint;
  testPoints: double;
begin
  for test := 1 to 10 do begin
    WriteStr(inputName, 'tests/input', test, '.txt');
    WriteStr(outputName, 'output', test, '.txt');
    ReadData(task, inputName);
    len := 0;
    for i := 1 to task.kMaxN do begin
      for j := 1 to task.kMaxM do begin
        solution[len].x := i;
        solution[len].y := j;
        solution[len].t := 0;
        Inc(len);
      end;
    end;
    testPoints := PrintSolution(task, solution, len, outputName);
    WriteLn('Test ', test, ': ', testPoints : 0 : 4, ' points');
    totalPoints := totalPoints + testPoints;
  end;
  WriteLn('Total: ', totalPoints : 0 : 4, ' points');
end.